boringtun-cli/src/main.rs

```diff
@@ -100,43 +111,54 @@ fn main() {
     let (sock1, sock2) = UnixDatagram::pair().unwrap();
     let _ = sock1.set_nonblocking(true);
 
-    let _guard;
+    let log = matches.value_of("log").unwrap();
 
     if background {
-        let log = matches.value_of("log").unwrap();
+        let mut daemonize =
+            Daemonize::new().working_directory(matches.value_of("work-dir").unwrap());
 
-        let log_file =
-            File::create(log).unwrap_or_else(|_| panic!("Could not create log file {}", log));
-
-        let (non_blocking, guard) = tracing_appender::non_blocking(log_file);
+        if let Some(pid_file) = matches.value_of("pid-file") {
+            daemonize = daemonize.pid_file(pid_file);
+        }
 
-        _guard = guard;
+        let outcome = daemonize.execute();
 
-        tracing_subscriber::fmt()
-            .with_max_level(log_level)
-            .with_writer(non_blocking)
-            .with_ansi(false)
-            .init();
+        match outcome {
+            Outcome::Parent(Ok(parent))=> {
 
-        let daemonize = Daemonize::new()
-            .working_directory("/tmp")
-            .exit_action(move || {
                 let mut b = [0u8; 1];
                 if sock2.recv(&mut b).is_ok() && b[0] == 1 {
-                    println!("BoringTun started successfully");
+                    eprintln!("BoringTun started successfully");
+                    exit(parent.first_child_exit_code);
                 } else {
                     eprintln!("BoringTun failed to start");
                     exit(1);
                 };
-            });
-
-        match daemonize.start() {
-            Ok(_) => tracing::info!("BoringTun started successfully"),
-            Err(e) => {
-                tracing::error!(error = ?e);
+            }
+            Outcome::Child(Ok(_)) => {
+                let log_file = OpenOptions::new()
+                    .create(true)
+                    .append(true)
+                    .open(log)
+                    .unwrap_or_else(move |_| panic!("Could not create log file {}", &log));
+
+                let (non_blocking, _guard) = tracing_appender::non_blocking(log_file);
```

Comment:
The `_guard` and thus log file will be dropped at the end of this branch. The `_guard` must be held at the top level in order for the log to keep working, i.e.

```rust
fn main() {
    ...
    let _guard;
    ...

                let (non_blocking, guard) = tracing_appender::non_blocking(log_file);
                _guard = guard;

    ...
```

this is basically the approach in the main branch

https://github.com/cloudflare/boringtun/blob/f672bb6c1e1e371240a8d151f15854687eb740bb/boringtun-cli/src/main.rs#L103-L113